### PR TITLE
Simplify configurations

### PR DESCRIPTION
**Pull request by andreievg**
Sorry no associated issue, it's me learning rust.

## Changes
* No explicit panics out of configuration.rs, always return ConfigError
* Move string to constants at the top of file (not sure if it adds anything, just a stylistic habit)
* Couldn't see the need to convert env var to Environment and then to $str, simplified
* Remove extra environment specification, wasn't sure why it's there

@wlthomson the last two, may need to be reinstated, but I couldn't figure out why they are need.

Would be nice to have conversion from error returned by current_dir to ConfigError, but they are from external crates (could make a common type, but thought map_err is enough)

### CODE REVIEW

**Review by wlthomson — APPROVED** [review]
Spent some time on this, because wanted to get it right (and was good learning opportunity for me).

After revisiting the code, I took some of @andreievg ideas and ran with them. I think we can make this even better with our own error types. This allows us to map between errors more easily, and also is more self-documenting wrt. what set of errors we are handling.

I've done a [PR with my suggested changes](#224).

File: src/util/configuration.rs [hunk 1]
```diff
@@ -1,26 +1,20 @@
...
-        .unwrap_or_else(|_| "local".into())
-        .parse()
-        .expect("Failed to parse APP_ENVIRONMENT");
+    let extra_file = std::env::var("APP_ENVIRONMENT").unwrap_or_else(|_| DEFAULT_EXTRA_FILE.into());
 
-    settings.merge(
-        config::File::from(configuration_directory.join(environment.as_str())).required(true),
-    )?;
-
-    settings.merge(config::Environment::with_prefix("app").separator("__"))?;
```
[the remark this answers is not in the document]
Comment (andreievg) [reply src/util/configuration.rs:23]: why have two overriding ways ? from base to (either local or production), then by obfuscated command line parameter ?

Ok to keep it as long as it's well documented (in code an in readme) `CONFIGURATION_ENVIRONMENT_PREFIX` and `CONFIGURATION_ENVIRONMENT_PREFIX` is not quite enough to realise what's going on 

> I think this is also not major, but worth keeping (if my opinion doesn't swat you, I stole this pretty much verbatim from Zero2Production 

The only way to decipher the functionality is to read the book or source code, there are 0 docs for config::Environment in the crate.

File: src/util/configuration.rs [hunk 2]
```diff
@@ -1,26 +1,20 @@
-use crate::util::environment::Environment;
 use crate::util::settings::Settings;
+use config::{Config, ConfigError, File};
 
-use std::borrow::Cow;
+static CONFIGURATION_FOLDER: &str = "configuration";
```
[the remark this answers is not in the document]
Comment (andreievg) [reply src/util/configuration.rs:4]: I never paid attention to this details lol, wonder why `cmd` has `mkdir`

**Comment by andreievg** [comment]
Thanks will for review, I've added some comments to your PR, wonder what we should do with this one, close or ?

**Comment by wlthomson** [comment]
> 
> 
> Thanks will for review, I've added some comments to your PR, wonder what we should do with this one, close or ?

Up to you! Mine is based off it, so there's no harm in merging it and then merging mine to follow-up. There are some issues with the documentation domain resetting merge, so it will be useful to have this to merge!

File: src/util/configuration.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (wlthomson) [reply src/util/configuration.rs:23]: Fair points! The idea with `base` is just DRYness, I believe. I believe it's no so much that `local` and `production` override `base`, but that `base` extracts what `local` and `production` have in common so it doesn't need to be repeated. I think the reasoning for the app env stuff is that `local` and `production` should be well-defined environments that are relatively static (i.e. they shouldn't change much). So, the idea is that `local` or `production` don't need to be directly mutated, instead the user can make minor temporary overrides where needed (e.g. just changing the port). 

Re. the documentation, I agree 100% that it needs to be clearly documented!

[the remark this answers is not in the document]
Comment (andreievg) [reply src/util/configuration.rs:23]: @wlthomson, cool, let's keep (in your PR), but please add comment when with_prefix is used, and readme

**Pull request merged**